### PR TITLE
[SPARK-41792][Shuffle] Fix DB update for push based shuffle when newer shuffle merge is received

### PR DESCRIPTION
**Pull request by mridulm**
### What changes were proposed in this pull request?
Incorrect merge id is removed from the DB when a newer shuffle merge id is received.

### Why are the changes needed?
Bug fix

### Does this PR introduce _any_ user-facing change?
No, fixes a corner case bug

### How was this patch tested?
Unit test updated

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @otterc, @Ngone51

**Comment by mridulm** [comment]
+CC @wankunde, this fixes the bug I had [described earlier](https://github.com/apache/spark/pull/37922#discussion_r990753031).

**Review by dongjoon-hyun — COMMENTED** [review]
According to the JIRA, is this for both Apache Spark 3.3.2 and 3.4.0, @mridulm ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. (BTW, nit, the Java-style 8-space indentation for continued lines looks a little strange to me. Since we used 2-space indentation even in Java, I thought we used 4-space for continued lines)

**Comment by dongjoon-hyun** [comment]
Oh, could you take a look at the failures? It looks like relevant.
```
[info] *** 2 TESTS FAILED ***
[error] Failed tests:
[error] 	org.apache.spark.network.yarn.YarnShuffleServiceWithRocksDBBackendSuite
[error] 	org.apache.spark.network.yarn.YarnShuffleServiceWithLevelDBBackendSuite
```

**Comment by wankunde** [comment]
+1, LGTM

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java [hunk 1]
```diff
@@ -227,15 +227,15 @@ AppShufflePartitionInfo getOrCreateAppShufflePartitionInfo(
...
             logger.info("{}: creating a new shuffle merge metadata since received " +
-                "shuffleMergeId is higher than latest shuffleMergeId {}",
-                appAttemptShuffleMergeId, latestShuffleMergeId);
+                "shuffleMergeId {} is higher than latest shuffleMergeId {}",
+                    currrentAppAttemptShuffleMergeId, shuffleMergeId, latestShuffleMergeId);
             submitCleanupTask(() ->
-                closeAndDeleteOutdatedPartitions(
-                    appAttemptShuffleMergeId, mergePartitionsInfo.shuffleMergePartitions));
+                closeAndDeleteOutdatedPartitions(currrentAppAttemptShuffleMergeId,
+                        mergePartitionsInfo.shuffleMergePartitions));
```
Comment (zhouyejoe) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:238]: Nit: 4 spaces for continued lines.

**Review by zhouyejoe — APPROVED** [review]
+1 LGTM

**Comment by yabola** [comment]
@mridulm 
I agree with you. The failed UT is as expected and we should modify the UT. 
My previous PR also do it https://github.com/apache/spark/pull/38560/files#r1031350719

Actually in my previous PR, I tried unify this deletion logic: delete `mergePartitionsInfo.shuffleMergePartitions` at the same time deleting the DB.
What more, I also want to unify the deleting the finalized partition information ( if higher mergeId come and merge partitions was finalized, we should clean finalized partitions, please see here https://github.com/apache/spark/pull/38560/files#diff-d544fbb952b61283b3d18ca10a5027528efc4f2f65047130da015ae7754c117fR502)

**Comment by mridulm** [comment]
Let me take a look, thanks @dongjoon-hyun.
Did not notice this failing in my local box ... weird.

**Comment by mridulm** [comment]
Agree with @yabola, the test is broken: I will fix the test to the expected output.

**Comment by mridulm** [comment]
@dongjoon-hyun:
> According to the JIRA, is this for both Apache Spark 3.3.2 and 3.4.0, @mridulm ?

This is only for master, I have updated the jira ...

**Comment by dongjoon-hyun** [comment]
Got it. Thank you, @mridulm

**Comment by mridulm** [comment]
Merging to master

**Comment by mridulm** [comment]
Thanks for the review @dongjoon-hyun, @wankunde, @zhouyejoe, @yabola :-)

**Pull request closed without merging**